### DISCUSSION TITLE
Proper kustomize syntax to replace the value of list element with condition

### DISCUSSION THREAD

**Discussion by kastl-ars** [kubernetes-sigs/kustomize#5663] [Q&A]
Given this [YAML file](https://github.com/grafana/intro-to-mltp/blob/main/k8s/mythical/mythical-beasts-deployment.yaml), what would be the proper syntax to patch the value of the `env` element that has `TRACING_COLLECTOR_HOST` as `name`?

This works, but hardcodes "use the second element in the list" (`.../env/1/...`):

```
- op: replace
  path: /spec/template/spec/containers/0/env/1/value
  value: newvalue
```

I find no syntax that kustomize accepts, that lets me define this. Is this possible?

The RFCs for JSON patches contain only very basic paths, unfortunately.

`yq` accepts something like this:
```
.spec.template.spec.containers.0.env[] | select(.name == "TRACING_COLLECTOR_HOST").value
```

Here are some of the things I tried with kustomize:

```
  path: /spec/template/spec/containers/0/env/(.name == 'TRACING_COLLECTOR_HOST')/value
  path: /spec/template/spec/containers/0/env[](.name == 'TRACING_COLLECTOR_HOST')/value
  path: /spec/template/spec/containers/0/env/(.name == 'TRACING_COLLECTOR_HOST')/value
  path: /spec/template/spec/containers/0/env/(name == 'TRACING_COLLECTOR_HOST')/value
  path: /spec/template/spec/containers/0/env/(/name == 'TRACING_COLLECTOR_HOST')/value
  path: /spec/template/spec/containers/0/env/name == 'TRACING_COLLECTOR_HOST'/value
...
```

**Comment by ChristianCiach** [comment]
This is simply not possible with `patchesJson6902`.

What is commonly done instead is to use the `test` operation to make sure that you are actually patching the expected element. Something like that:

```yaml
- op: test
  path: /spec/template/spec/containers/0/env/1/name
  value: TRACING_COLLECTOR_HOST
- op: replace
  path: /spec/template/spec/containers/0/env/1/value
  value: newvalue
```

You should probably also add a `test` for `/spec/template/spec/containers/0/name` to make sure that you are patching the expected container.

**Reply by ChristianCiach** [reply]
If/when https://github.com/kubernetes-sigs/kustomize/pull/5630 gets merged, you could probably use the `replacement` feature to patch this element, because the `replacement` feature supports a mightier syntax for selecting the target element.

**Comment by sda399** [comment]
you can use replacements with a local configmap (config.kubernetes.io/local-config: true)
